firebase-installations/src/main/java/com/google/firebase/installations/remote/ResponseHandler.java

```diff
@@ -0,0 +1,87 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.firebase.installations.remote;
+
+import androidx.annotation.GuardedBy;
+import com.google.firebase.installations.Utils;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * The {@link ResponseHandler} class calculates the next allowed request time. Also, decides whether
+ * a network request to FIS servers is allowed to execute.
+ *
+ * @hide
+ */
+class ResponseHandler {
+  private static final long BACKOFF_TIME_24H_IN_MILLIS = TimeUnit.HOURS.toMillis(24);
+  private static final long BACKOFF_TIME_30_MINS_IN_MILLIS = TimeUnit.MINUTES.toMillis(30);
+  private final Utils utils;
+
+  @GuardedBy("this")
+  private long nextAllowedRequestTime = Long.MAX_VALUE;
+
+  @GuardedBy("this")
+  private int attemptCount = 0;
+
+  ResponseHandler(Utils utils) {
+    this.utils = utils;
+  }
+
+  public synchronized void setNextAllowedRequestTime(int responseCode) {
+    if (responseCode == 200) {
+      resetAllowedRequestTime();
+      return;
+    }
+    attemptCount++;
+    long backOffTime = getBackoffTime(responseCode);
+    nextAllowedRequestTime = utils.currentTimeInMillis() + backOffTime;
+  }
+
+  private synchronized void resetAllowedRequestTime() {
+    nextAllowedRequestTime = Long.MAX_VALUE;
+    attemptCount = 0;
+  }
+
+  private synchronized long getBackoffTime(int responseCode) {
+    // Fixed 24 hours silence period for non-retryable errors. Read more: b/160751425.
+    if (isNonRetryableError(responseCode)) {
+      return BACKOFF_TIME_24H_IN_MILLIS;
+    }
+    // Quickly increasing dynamically configured back-off strategy for Retryable errors. Read more:
+    // https://cloud.google.com/storage/docs/exponential-backoff.
+    return (long)
+        Math.min(
+            Math.pow(2, attemptCount) + utils.getRandomMillis(), BACKOFF_TIME_30_MINS_IN_MILLIS);
+  }
+
+  // Response codes classified as non-retryable for FIS API, all other response codes will be
+  // treated as retryable-errors. Read more: go/fis-api-error-code-classification.
+  private static boolean isNonRetryableError(int responseCode) {
+    return responseCode == 400 || responseCode == 403;
+  }
+
+  public boolean isRequestAllowed() {
+    long currentTime = utils.currentTimeInMillis();
+    boolean isRequestAllowed =
+        attemptCount == 0 || currentTime > nextAllowedRequestTime;
+
+    if (currentTime > nextAllowedRequestTime) {
+      synchronized (this) {
+        nextAllowedRequestTime = Long.MAX_VALUE;
```

Comment:
I wonder what's the intended used of  `isRequestAllowed` method? Will this line lead to the method returning `false` until `setNextAllowedRequestTime()` method is called?

Reply:
How about instead we set `attemptCount = 0`.

Reply:
@maksymmalyhin 
isRequestAllowed will be called before making a network call to FIS Servers. This method will return true if its a first call i.e attemptCount = 0 or if currentTime is greater than the set nextRequestTime.

The way I imagined plugging these methods was `setNextAllowedRequestTime()` will be called on every responseCode resulting in correctly setting `attemptNumber` & `nextRequestTime`.

But I guess, we can do without explicitly resetting `nextRequestTime`. WDYT?

Reply:
@andirayo I don't think setting it attemptCount = 0 will be correct.

Let's say FIS servers respond 429/ 5xx for consecutive requests.

`Response1 code 500 : At Duration T1 with attemptCount = 1, nextRequestTime = T2`

Next Request to communicate with FIS servers is at T3 > T2. Since this is true, we allow the request to proceed. But resetting attemptCount = 0 will beat the purpose of exponential backoff. 

If this request also fails, attemptCount should be 2:
`Response2 code 500: At  Duration T3 with attemptCount = 2, nextRequestTime = T4`

Please correct me if I have misunderstood something here. 

Thanks.

Reply:
IMO this logic places unnecessary restrictions on the client and makes some potential use cases complicated e.g. if the client invokes `isRequestAllowed` and then doesn't send a request because of another condition, there is no a straightforward way to call `isRequestAllowed` and get expected result. I would prefer keep returning `true` until a signal from the client (via `setNextAllowedRequestTime()`) is received. So, basically, I would just remove lines 80-83. WDYT?

Reply:
Agreed. I realized that there is no need to reset the nextRequestTime here. Changed it.

Thanks for the reveiw.